src/node_api.cc

```diff
@@ -0,0 +1,2478 @@
+                               napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  v8::Isolate* isolate = v8impl::V8IsolateFromJsEnv(env);
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+
+  v8::Local<v8::Object> obj;
+  CHECK_TO_OBJECT(context, obj, object);
+
+  v8::Local<v8::Value> val = obj->GetPrototype();
+  *result = v8impl::JsValueFromV8LocalValue(val);
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_object(napi_env env, napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(
+      v8::Object::New(v8impl::V8IsolateFromJsEnv(env)));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_array(napi_env env, napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(
+      v8::Array::New(v8impl::V8IsolateFromJsEnv(env)));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_array_with_length(napi_env env,
+                                          size_t length,
+                                          napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(
+      v8::Array::New(v8impl::V8IsolateFromJsEnv(env), length));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_string_utf8(napi_env env,
+                                    const char* str,
+                                    size_t length,
+                                    napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  auto isolate = v8impl::V8IsolateFromJsEnv(env);
+  v8::Local<v8::String> s;
+  CHECK_NEW_FROM_UTF8_LEN(isolate, s, str, length);
+
+  *result = v8impl::JsValueFromV8LocalValue(s);
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_string_utf16(napi_env env,
+                                     const char16_t* str,
+                                     size_t length,
+                                     napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  auto isolate = v8impl::V8IsolateFromJsEnv(env);
+  auto str_maybe =
+      v8::String::NewFromTwoByte(isolate,
+                                 reinterpret_cast<const uint16_t*>(str),
+                                 v8::NewStringType::kInternalized,
+                                 length);
+  CHECK_MAYBE_EMPTY(str_maybe, napi_generic_failure);
+
+  *result = v8impl::JsValueFromV8LocalValue(str_maybe.ToLocalChecked());
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_number(napi_env env,
+                               double value,
+                               napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(
+      v8::Number::New(v8impl::V8IsolateFromJsEnv(env), value));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_get_boolean(napi_env env, bool value, napi_value* result) {
```

Comment:
The thinking was that the `undefined`, `null`, `true`, and `false` values are conceptually singleton values: you don't ever actually create an instance of one of them you just get a handle. I guess whether that's actually true or not depends on the VM implementation, since they are not cached at the N-API layer. Personally I'd be fine with either name, and I don't think anyone feels strongly about it.